### PR TITLE
length check fixes

### PR DESCRIPTION
**Pull request by akallabeth**

### CODE REVIEW

**Comment by akallabeth** [comment]
@freerdp-bot test

File: libfreerdp/codec/zgfx.c [hunk 1]
```diff
@@ -227,7 +227,7 @@ static BOOL zgfx_decompress_segment(ZGFX_CONTEXT* zgfx, wStream* stream, size_t
 	BYTE* pbSegment;
 	size_t cbSegment;
 
-	if (!zgfx || !stream)
+	if (!zgfx || !stream || (segmentSize < 2))
```
[the remark this answers is not in the document]
Comment (akallabeth) [reply libfreerdp/codec/zgfx.c:230]: true

File: channels/drive/client/drive_file.c [hunk 2]
```diff
@@ -73,71 +77,100 @@ static void drive_file_fix_path(WCHAR* path)
...
+	if (!drive_file_fix_path(fullpath, length))
+		goto fail;
+
+	buffer = calloc(length + 1, sizeof(WCHAR));
+	if (!buffer)
+		goto fail;
+
+	/* Ensure the path does not contain sequences like '..' */
+	const WCHAR dotdot[] = { '.', '.', '\0' };
+	if (_wcsstr(&buffer[base_path_length], dotdot))
```
Comment (mfleisz) [inline channels/drive/client/drive_file.c:126]: `buffer` is never initialized (apart from being zeroed)

File: winpr/libwinpr/crt/string.c [hunk 3]
```diff
@@ -144,6 +165,25 @@ size_t _wcsnlen(const WCHAR* str, size_t max)
...
+WCHAR* _wcsstr(const WCHAR* str, const WCHAR* strSearch)
+{
+	WINPR_ASSERT(str);
+	WINPR_ASSERT(strSearch);
+
+	if (strSearch[0] == '\0')
+		return str;
+
+	const size_t searchLen = _wcslen(strSearch);
+	while (*str)
```
Comment (mfleisz) [inline winpr/libwinpr/crt/string.c:179]: This will loop forever if `str` does not start with `strSearch`

File: winpr/libwinpr/crt/string.c [hunk 4]
```diff
@@ -98,17 +99,39 @@ int _strnicmp(const char* string1, const char* string2, size_t count)
 
 int _wcscmp(const WCHAR* string1, const WCHAR* string2)
 {
-	WCHAR value1, value2;
+	WINPR_ASSERT(string1);
+	WINPR_ASSERT(string2);
 
-	while (*string1 && (*string1 == *string2))
+	while (*string1 && (*string2))
```
Comment (mfleisz) [inline winpr/libwinpr/crt/string.c:105]: Why not just add a check for `*string2` to the old loop. With these changes we now check each string twice for `\0` each iteration

File: winpr/libwinpr/crt/string.c (hunk 3 above)
Comment (akallabeth) [reply winpr/libwinpr/crt/string.c:179]: ok, insert coffee

File: winpr/libwinpr/crt/string.c (hunk 4 above)
Comment (akallabeth) [reply winpr/libwinpr/crt/string.c]: replaced with a `while(TRUE)` and the `\0` checks below.

File: channels/drive/client/drive_file.c [hunk 5]
```diff
@@ -73,71 +77,100 @@ static void drive_file_fix_path(WCHAR* path)
...
+		goto fail;
+
+	CopyMemory(fullpath, base_path, base_path_length * sizeof(WCHAR));
+	if (path)
+		CopyMemory(&fullpath[base_path_length], path, PathWCharLength * sizeof(WCHAR));
+
+	if (!drive_file_fix_path(fullpath, length))
+		goto fail;
+
+	buffer = calloc(length + 1, sizeof(WCHAR));
```
Comment (mfleisz) [inline channels/drive/client/drive_file.c:120]: `buffer` seems to be unused

Comment (akallabeth) [reply channels/drive/client/drive_file.c]: true, now `fullpath` does the same.

**Review by mfleisz — APPROVED** [review]

**Pull request merged**